### PR TITLE
Added support for owner-delegated grant

### PR DESCRIPTION
**Pull request by thehenrytsai**
1. Added support for owner-delegated grant
1. Added missing checks for author-deleted grant:
   1. Disallowed invocation of non-delegated grant as delegated grant
   1. Disallowed mismatching of grant ID referenced and CID of actual given grant
2. Added numerous tests to catch up on missing scenario tests for delegated grants in general

### CODE REVIEW

File: tests/features/author-delegated-grant.spec.ts [hunk 1]
```diff
@@ -62,7 +63,82 @@ export function testDelegatedGrantScenarios(): void {
...
+          grantedBy   : alice.did,
+          grantedTo   : bob.did,
+          grantedFor  : alice.did,
+          scope,
+          signer      : Jws.createSigner(alice)
+        });
+
+        // Bob creates a chat message invoking the delegated grant (ID) but does not include the author-delegated grant (we remove it below)
+        const recordsWrite = await RecordsWrite.create({
+          signer         : Jws.createSigner(alice),
```
Comment (LiranCohen) [inline tests/features/author-delegated-grant.spec.ts:90]: Sanity: should the signer be bob?

**Review by LiranCohen — DISMISSED** [review]
🐐 as always. Just a few small questions/nits.

File: src/interfaces/records-write.ts [hunk 2]
```diff
@@ -212,6 +219,19 @@ export class RecordsWrite implements MessageInterface<RecordsWriteMessage> {
     return Message.getSigner(this._message);
   }
 
+  /**
+   * Gets the signer of owner signature.
+   * This is not to be confused with the logical owner of the message.
+   */
+  public get ownerSignatureSigner(): string | undefined {
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/interfaces/records-write.ts:226]: Good call, added more documentation to both locations.

File: src/interfaces/records-write.ts [hunk 3]
```diff
@@ -531,6 +564,31 @@ export class RecordsWrite implements MessageInterface<RecordsWriteMessage> {
...
+  /**
+   * Signs the `RecordsWrite` as the DWN owner-delegate.
+   * This is used when a DWN owner-delegate wants to retain a copy of a message that the owner did not author.
+   * NOTE: requires the `RecordsWrite` to already have the author's signature.
+   */
+  public async signAsOwnerDelegate(signer: Signer, delegatedGrant: DelegatedGrantMessage): Promise<void> {
+    if (this._author === undefined) {
+      throw new DwnError(
+        DwnErrorCode.RecordsWriteSignAsOwnerUnknownAuthor,
+        'Unable to sign as owner if without message signature because owner needs to sign over `recordId` which depends on author DID.');
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/interfaces/records-write.ts:576]: Thanks for the good catch, and noticed missing code coverage! Will add now.

File: src/interfaces/records-write.ts [hunk 4]
```diff
@@ -531,6 +564,31 @@ export class RecordsWrite implements MessageInterface<RecordsWriteMessage> {
...
+
+    const descriptor = this._message.descriptor;
+    const ownerSignature = await Message.createSignature(descriptor, signer, { delegatedGrantId });
+
+    this._message.authorization!.ownerSignature = ownerSignature;
+    this._message.authorization!.ownerDelegatedGrant = delegatedGrant;
+
+    this._ownerSignaturePayload = Jws.decodePlainObjectPayload(ownerSignature);
+    this._owner = Jws.getSignerDid(delegatedGrant.authorization.signature.signatures[0]);
+    ;
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/interfaces/records-write.ts:589]: Removed.

File: tests/features/author-delegated-grant.spec.ts [hunk 5]
```diff
@@ -579,19 +655,80 @@ export function testDelegatedGrantScenarios(): void {
       expect(bobRecordsQueryReply2.entries?.length).to.equal(0);
     });
 
-    xit('should not allow entity using a non-delegated grant as a delegated grant to invoke write', async () => {
+    it('should not allow entity using a non-delegated grant as an author-delegated grant to invoke write', async () => {
+      // scenario:
+      // 1. Bob has the message protocol installed
+      // 2. Alice creates a delegated grant for device X
+      // 3. Verify that device X cannot write a `RecordsWrite` message to Bob's DWN as Alice using the non-delegated grant
+      // 4. Verify the message by device X got written to Bob's DWN, AND Alice is the logical author
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply tests/features/author-delegated-grant.spec.ts:663]: Thanks for the keen eyes! Fixed.

File: tests/features/author-delegated-grant.spec.ts [hunk 6]
```diff
@@ -579,19 +655,80 @@ export function testDelegatedGrantScenarios(): void {
...
+      const protocolDefinition = messageProtocolDefinition;
+      const protocol = protocolDefinition.protocol;
+      const protocolsConfig = await TestDataGenerator.generateProtocolsConfigure({
+        author: bob,
+        protocolDefinition
+      });
+      const protocolConfigureReply = await dwn.processMessage(bob.did, protocolsConfig.message);
+      expect(protocolConfigureReply.status.code).to.equal(202);
+
+      // 2. Alice creates a delegated grant for device X
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply tests/features/author-delegated-grant.spec.ts:678]: Fixed.

File: tests/features/author-delegated-grant.spec.ts [hunk 7]
```diff
@@ -62,7 +63,82 @@ export function testDelegatedGrantScenarios(): void {
...
+          grantedTo   : bob.did,
+          grantedFor  : alice.did,
+          scope,
+          signer      : Jws.createSigner(alice)
+        });
+
+        // Bob attempts to sign as Alice by including an author-delegated grant
+        // but does not reference the grant ID in author signature (we remove it below)
+        const recordsWrite = await RecordsWrite.create({
+          signer         : Jws.createSigner(alice),
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply tests/features/author-delegated-grant.spec.ts:126]: Should be fixed for sure to match the scenario description even though it doesn't actually matter.

File: tests/features/author-delegated-grant.spec.ts (hunk 1 above)
Comment (thehenrytsai) [reply tests/features/author-delegated-grant.spec.ts:90]: You are right, should be fixed like the other test!

**Review by LiranCohen — APPROVED** [review]
🚀 🚀 🚀

**Pull request merged**